cpp/include/Ice/SlicedData.h

```diff
@@ -59,6 +60,14 @@ namespace Ice
               hasOptionalMembers(hasOptionalMembers),
               isLastSlice(isLastSlice)
         {
+            if (compactId == -1)
+            {
+                assert(typeId.empty());
+            }
+            else
+            {
+                assert(!typeId.empty());
+            }
```

Comment:
I find a bit odd to have if/else that just contain an assert, what about:

```suggestion
            assert((compactId == -1 && typeId.empty()) || (compactId != -1 && !typeId.empty()));
```

Reply:
I applied Joe's suggestion.

Reply:
On second thought, I reverted to an `if`: I prefer two asserts.

Also, in this particular code, we need to use `this->typeId` because `typeId` is the moved-out argument.